In [1]:
import pandas as pd

In [2]:
path = '../raw_data/triples.csv'
df = pd.read_csv(path, header=0)

In [3]:
df.head()

head tail               relation label
0   关羽   刘备  younger_sworn_brother    义弟
1   张飞   刘备  younger_sworn_brother    义弟
2   关羽   张飞    elder_sworn_brother    义兄
3   张苞   张飞                    son    儿子
4   关兴   关羽                    son    儿子

超昂 应该是曹昂，之后把所有内容都更改一下

145   超昂   刘氏      son    儿子

In [70]:
df.at[145, 'head']

'超昂'

In [71]:
df.at[145, 'head'] = '曹昂'
df.at[145, 'head']

'曹昂'

In [113]:
head_identicle_people = set(df['head'].tolist())
tail_identicle_people = set(df['tail'].tolist())

In [114]:
all_identicle_people = set.union(head_identicle_people, tail_identicle_people)

In [115]:
print(all_identicle_people)
print(len(all_identicle_people))

{'孙坚', '荀攸', '沙摩柯', '夏侯楙', '徐晃', '孙氏', '黄承彦', '甘氏', '徐盛', '关兴', '庞德', '吕伯奢', '刘辩', '魏延', '韩遂', '程普', '太史慈', '诸葛亮', '丁原', '袁谭', '张角', '曹纯', '刘胜', '甘宁', '张辽', '蔡瑁', '公孙越', '张梁', '曹嵩', '张宝', '司马懿', '鲁肃', '邓艾', '夏侯渊', '糜竺', '邹氏', '程昱', '刘备', '司马炎', '韩当', '祝融', '刘氏', '关平', '清河公主', '吕蒙', '庞德公', '曹丕', '曹昂', '张苞', '于禁', '张昭', '袁术', '马腾', '诸葛瞻', '蒯越', '孙策', '钟会', '孙韶', '法正', '司马师', '袁熙', '黄忠', '贾诩', '孟获', '刘表', '曹操', '张郃', '蒋钦', '卢植', '刘协', '貂蝉', '曹爽', '文聘', '吴国太', '夏侯淳', '曹真', '王允', '陆逊', '凌统', '袁尚', '荀彧', '高顺', '大乔', '周泰', '关羽', '孙尚香', '陈宫', '许褚', '吕布', '刘禅', '黄月英', '董卓', '庞统', '马超', '袁绍', '典韦', '郭嘉', '张飞', '徐庶', '乐进', '诸葛瑾', '张绣', '公孙瓒', '马良', '丁奉', '刘启', '卞氏', '赵云', '孙权', '黄盖', '司马昭', '曹仁', '曹植', '黄祖', '蔡氏', '糜芳', '周瑜', '姜维', '糜氏', '蒋琬', '马谡', '小乔'}
122


In [5]:
def find_relation(name):
    p1 = df.loc[df['head'] == name]
    p2 = df.loc[df['tail'] == name]
    print(p1)
    print(p2)

In [75]:
find_relation( '曹昂' )

    head tail relation label
143   曹昂   曹操      son    儿子
145   曹昂   刘氏      son    儿子
Empty DataFrame
Columns: [head, tail, relation, label]
Index: []


In [110]:
find_relation( '超昂' )

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
Empty DataFrame
Columns: [head, tail, relation, label]
Index: []


夏侯惇被错写为了夏侯淳，更改所有的位置。

In [8]:
find_relation('夏侯淳')

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
    head tail      relation label
114  夏侯渊  夏侯淳  son_of_uncle    堂弟
115   曹操  夏侯淳     same_clan    同族
116   曹操  夏侯淳          lord    主公


In [9]:
df.at[114, 'tail'] = '夏侯惇'
df.at[115, 'tail'] = '夏侯惇'
df.at[116, 'tail'] = '夏侯惇'

In [10]:
find_relation('夏侯淳')

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
Empty DataFrame
Columns: [head, tail, relation, label]
Index: []


In [11]:
find_relation('夏侯惇')

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
    head tail      relation label
114  夏侯渊  夏侯惇  son_of_uncle    堂弟
115   曹操  夏侯惇     same_clan    同族
116   曹操  夏侯惇          lord    主公


一共有122个不同的武将。这里我们将这些武将分为4大势力，蜀国，魏国，吴国，群雄

魏国： '荀攸' '徐晃' '庞德'  '曹操'  '张郃' '荀彧' '司马昭' '曹仁' '曹植' '典韦' '郭嘉' '乐进' '张辽' '夏侯淳' '曹真' '曹爽' '曹纯' '司马炎' '司马懿' '邓艾' '夏侯渊' '贾诩' '许褚' '司马师' '夏侯楙' '蔡瑁' '曹嵩' '邹氏' '程昱' '刘氏' '清河公主' '于禁' '蒯越' '钟会' '文聘' '曹昂'
 
蜀国：'关兴' '魏延' '诸葛亮' '诸葛瞻' '法正' '黄忠' '关羽' '赵云' '姜维' '张飞' '徐庶' '马良' '诸葛瑾'  '糜芳'  '刘禅' '黄月英' '庞统' '马超' '祝融'  '黄承彦' '沙摩柯' '甘氏' '糜竺' '刘备' '关平' '张苞' '糜氏' '蒋琬' '马谡' 
 
吴国：'孙坚' '太史慈' '孙策' '大乔' '周泰' '孙权' '黄盖' '周瑜' '甘宁' '吴国太'  '吕蒙' '曹丕' '曹昂' '鲁肃'  '孙尚香' '陆逊' '程普' '徐盛' '孙氏'  '韩当' '张昭' '孙韶' '蒋钦' '凌统' '丁奉' '小乔'
 
 
群雄：'袁术' '马腾' '陈宫' '公孙瓒'  '吕布'  '袁绍' '董卓' '孟获' '貂蝉' '吕伯奢' '刘辩' '韩遂' '丁原' '张角' '袁谭' '刘胜' '公孙越' '张梁' '张宝'  '庞德公' '袁熙'  '刘表'  '卢植' '刘协' '王允' '袁尚' '高顺' '张绣' '刘启' '卞氏' '黄祖' '蔡氏'

In [15]:
def string2list(country):
    """convert name string to name list"""
    country = country.replace("'", '')
    return country.split()

In [16]:
wei = "'荀攸' '徐晃' '庞德'  '曹操'  '张郃' '荀彧' '司马昭' '曹仁' '曹植' '典韦' '郭嘉' '乐进' '张辽' '夏侯淳' '曹真' '曹爽' '曹纯' '司马炎' '司马懿' '邓艾' '夏侯渊' '贾诩' '许褚' '司马师' '夏侯楙' '蔡瑁' '曹嵩' '邹氏' '程昱' '刘氏' '清河公主' '于禁' '蒯越' '钟会' '文聘' '曹昂'"
shu = "'关兴' '魏延' '诸葛亮' '诸葛瞻' '法正' '黄忠' '关羽' '赵云' '姜维' '张飞' '徐庶' '马良' '诸葛瑾'  '糜芳'  '刘禅' '黄月英' '庞统' '马超' '祝融'  '黄承彦' '沙摩柯' '甘氏' '糜竺' '刘备' '关平' '张苞' '糜氏' '蒋琬' '马谡'"
wu = "'孙坚' '太史慈' '孙策' '大乔' '周泰' '孙权' '黄盖' '周瑜' '甘宁' '吴国太'  '吕蒙' '曹丕' '曹昂' '鲁肃'  '孙尚香' '陆逊' '程普' '徐盛' '孙氏'  '韩当' '张昭' '孙韶' '蒋钦' '凌统' '丁奉' '小乔'"
others = "'袁术' '马腾' '陈宫' '公孙瓒'  '吕布'  '袁绍' '董卓' '孟获' '貂蝉' '吕伯奢' '刘辩' '韩遂' '丁原' '张角' '袁谭' '刘胜' '公孙越' '张梁' '张宝'  '庞德公' '袁熙'  '刘表'  '卢植' '刘协' '王允' '袁尚' '高顺' '张绣' '刘启' '卞氏' '黄祖' '蔡氏'"

In [17]:
wei = string2list(wei)
shu = string2list(shu)
wu = string2list(wu)
others = string2list(others)

In [18]:
len(wei) + len(shu) + len(wu) + len(others)

123

到此为止得到了保存有三个势力的

In [19]:
four_camp = wei + shu + wu + others 
len(four_camp)

123

four_camp的总人数是123人，说明有重复的一人。下面写个函数，来找到是哪个人加入了两个势力。

In [20]:
for people in four_camp:
    count = four_camp.count(people)
    if count == 2:
        print(people)

曹昂
曹昂


好的，发现是曹昂加入了吴国阵营。。下面重新设定一下人物列表。为了方便给之后的pandas添加新的一列，这里用dict来保存所有人的配属：

In [21]:
wei = "'荀攸' '徐晃' '庞德'  '曹操'  '张郃' '荀彧' '司马昭' '曹仁' '曹植' '典韦' '郭嘉' '乐进' '张辽' '夏侯惇' '曹真' '曹爽' '曹纯' '司马炎' '司马懿' '邓艾' '夏侯渊' '贾诩' '许褚' '司马师' '夏侯楙' '蔡瑁' '曹嵩' '邹氏' '程昱' '刘氏' '清河公主' '于禁' '蒯越' '钟会' '文聘' '曹昂'"
shu = "'关兴' '魏延' '诸葛亮' '诸葛瞻' '法正' '黄忠' '关羽' '赵云' '姜维' '张飞' '徐庶' '马良' '诸葛瑾'  '糜芳'  '刘禅' '黄月英' '庞统' '马超' '祝融'  '黄承彦' '沙摩柯' '甘氏' '糜竺' '刘备' '关平' '张苞' '糜氏' '蒋琬' '马谡'"
wu = "'孙坚' '太史慈' '孙策' '大乔' '周泰' '孙权' '黄盖' '周瑜' '甘宁' '吴国太'  '吕蒙' '曹丕' '鲁肃'  '孙尚香' '陆逊' '程普' '徐盛' '孙氏'  '韩当' '张昭' '孙韶' '蒋钦' '凌统' '丁奉' '小乔'"
others = "'袁术' '马腾' '陈宫' '公孙瓒'  '吕布'  '袁绍' '董卓' '孟获' '貂蝉' '吕伯奢' '刘辩' '韩遂' '丁原' '张角' '袁谭' '刘胜' '公孙越' '张梁' '张宝'  '庞德公' '袁熙'  '刘表'  '卢植' '刘协' '王允' '袁尚' '高顺' '张绣' '刘启' '卞氏' '黄祖' '蔡氏'"

In [24]:
def assign_group(people_group, group, group_name=''):
    group = group.replace("'", '')
    group = group.split()
    for people in group:
        people_group[people] = group_name
    return people_group

In [25]:
people_group = {}
people_group = assign_group(people_group, wei, group_name='魏国')
people_group = assign_group(people_group, shu, group_name='蜀国')
people_group = assign_group(people_group, wu, group_name='吴国')
people_group = assign_group(people_group, others, group_name='群雄')

In [142]:
people_group

{'丁原': '群雄',
 '丁奉': '吴国',
 '乐进': '魏国',
 '于禁': '魏国',
 '公孙瓒': '群雄',
 '公孙越': '群雄',
 '关兴': '蜀国',
 '关平': '蜀国',
 '关羽': '蜀国',
 '典韦': '魏国',
 '凌统': '吴国',
 '刘协': '群雄',
 '刘启': '群雄',
 '刘备': '蜀国',
 '刘氏': '魏国',
 '刘禅': '蜀国',
 '刘胜': '群雄',
 '刘表': '群雄',
 '刘辩': '群雄',
 '卞氏': '群雄',
 '卢植': '群雄',
 '司马师': '魏国',
 '司马懿': '魏国',
 '司马昭': '魏国',
 '司马炎': '魏国',
 '吕伯奢': '群雄',
 '吕布': '群雄',
 '吕蒙': '吴国',
 '吴国太': '吴国',
 '周泰': '吴国',
 '周瑜': '吴国',
 '夏侯楙': '魏国',
 '夏侯淳': '魏国',
 '夏侯渊': '魏国',
 '大乔': '吴国',
 '太史慈': '吴国',
 '姜维': '蜀国',
 '孙坚': '吴国',
 '孙尚香': '吴国',
 '孙权': '吴国',
 '孙氏': '吴国',
 '孙策': '吴国',
 '孙韶': '吴国',
 '孟获': '群雄',
 '小乔': '吴国',
 '庞德': '魏国',
 '庞德公': '群雄',
 '庞统': '蜀国',
 '张宝': '群雄',
 '张昭': '吴国',
 '张梁': '群雄',
 '张绣': '群雄',
 '张苞': '蜀国',
 '张角': '群雄',
 '张辽': '魏国',
 '张郃': '魏国',
 '张飞': '蜀国',
 '徐庶': '蜀国',
 '徐晃': '魏国',
 '徐盛': '吴国',
 '文聘': '魏国',
 '曹丕': '吴国',
 '曹仁': '魏国',
 '曹嵩': '魏国',
 '曹操': '魏国',
 '曹昂': '魏国',
 '曹植': '魏国',
 '曹爽': '魏国',
 '曹真': '魏国',
 '曹纯': '魏国',
 '沙摩柯': '蜀国',
 '法正': '蜀国',
 '清河公主': '魏国',
 '王允': '群雄',
 '甘宁': '吴国',
 '甘氏': '

In [143]:
len(people_group)

122

人物数目一致。接下来创建两列，给每行的两个任务添加所属的阵营。直接使用map

In [155]:
df['head_group'] = df['head'].map(lambda x:people_group[x])
df['tail_group'] = df['tail'].map(lambda x:people_group[x])

In [156]:
df

head tail label head_group tail_group
0     关羽   刘备    义弟         蜀国         蜀国
1     张飞   刘备    义弟         蜀国         蜀国
2     关羽   张飞    义兄         蜀国         蜀国
3     张苞   张飞    儿子         蜀国         蜀国
4     关兴   关羽    儿子         蜀国         蜀国
5     关平   张苞    结拜         蜀国         蜀国
6     关平   关羽    义子         蜀国         蜀国
7     卢植   刘备    师傅         群雄         蜀国
8    公孙瓒   刘备    朋友         群雄         蜀国
9     甘氏   刘备     妻         蜀国         蜀国
10    刘禅   甘氏    儿子         蜀国         蜀国
11   诸葛瞻   刘禅    女婿         蜀国         蜀国
12   诸葛瞻  诸葛亮    儿子         蜀国         蜀国
13   诸葛亮   姜维    师傅         蜀国         蜀国
14    刘备  诸葛亮    主公         蜀国         蜀国
15   黄月英  诸葛亮     妻         蜀国         蜀国
16   黄承彦  黄月英    父亲         蜀国         蜀国
17   诸葛瑾  诸葛亮    哥哥         蜀国         蜀国
18   公孙越  公孙瓒    弟弟         群雄         群雄
19    刘备   马超    主公         蜀国         蜀国
20    马腾   马超    父亲         群雄         蜀国
21   公孙越  公孙瓒    弟弟         群雄         群雄
22    韩遂   马腾    结拜         群雄         群雄
23    马腾   马超    父亲         群雄         蜀国
24    刘备   徐庶    主公         蜀国         蜀国
25    曹操   徐庶    主公         魏国         蜀国
26    刘胜   刘启    儿子         群雄         群雄
27    刘辩   刘启    同族         群雄         群雄
28    甘氏   刘备     妻         蜀国         蜀国
29    刘禅   甘氏    儿子         蜀国         蜀国
..   ...  ...   ...        ...        ...
123   曹操   乐进    主公         魏国         魏国
124   曹操   张郃    主公         魏国         魏国
125   曹操   许褚    主公         魏国         魏国
126   曹操   典韦    主公         魏国         魏国
127   曹操   荀彧    主公         魏国         魏国
128   曹操   荀攸    主公         魏国         魏国
129   荀彧   荀攸    叔叔         魏国         魏国
130   曹操   贾诩    主公         魏国         魏国
131   曹操  司马懿    主公         魏国         魏国
132   曹操   程昱    主公         魏国         魏国
133   曹操   于禁    主公         魏国         魏国
134   曹操   邓艾    主公         魏国         魏国
135   曹操   钟会    主公         魏国         魏国
136   曹操   庞德    主公         魏国         魏国
137  司马师  司马懿    儿子         魏国         魏国
138  司马昭  司马师    弟弟         魏国         魏国
139  司马昭  司马懿    儿子         魏国         魏国
140  司马炎  司马昭    儿子         魏国         魏国
141   曹仁   曹操    同族         魏国         魏国
142   曹纯   曹仁    弟弟         魏国         魏国
143   曹昂   曹操    儿子         魏国         魏国
144   刘氏   曹操    妻子         魏国         魏国
145   曹昂   刘氏    儿子         魏国         魏国
146   卞氏   曹操    妻子         群雄         魏国
147   曹丕   卞氏    儿子         吴国         群雄
148   曹植   曹丕    弟弟         魏国         吴国
149   马谡  诸葛亮    同事         蜀国         蜀国
150   马谡   刘备     臣         蜀国         蜀国
151   孙坚   孙权    父亲         吴国         吴国
152  吴国太   孙权    母亲         吴国         吴国

[153 rows x 5 columns]

最后把relation这一列去掉就行了

In [152]:
df = df.drop(['relation'], axis=1)

In [153]:
df

head tail label head_group tail_group
0     关羽   刘备    义弟         蜀国         吴国
1     张飞   刘备    义弟         蜀国         吴国
2     关羽   张飞    义兄         蜀国         吴国
3     张苞   张飞    儿子         蜀国         吴国
4     关兴   关羽    儿子         蜀国         吴国
5     关平   张苞    结拜         蜀国         吴国
6     关平   关羽    义子         蜀国         吴国
7     卢植   刘备    师傅         群雄         吴国
8    公孙瓒   刘备    朋友         群雄         吴国
9     甘氏   刘备     妻         蜀国         吴国
10    刘禅   甘氏    儿子         蜀国         吴国
11   诸葛瞻   刘禅    女婿         蜀国         吴国
12   诸葛瞻  诸葛亮    儿子         蜀国         吴国
13   诸葛亮   姜维    师傅         蜀国         吴国
14    刘备  诸葛亮    主公         蜀国         吴国
15   黄月英  诸葛亮     妻         蜀国         吴国
16   黄承彦  黄月英    父亲         蜀国         吴国
17   诸葛瑾  诸葛亮    哥哥         蜀国         吴国
18   公孙越  公孙瓒    弟弟         群雄         吴国
19    刘备   马超    主公         蜀国         吴国
20    马腾   马超    父亲         群雄         吴国
21   公孙越  公孙瓒    弟弟         群雄         吴国
22    韩遂   马腾    结拜         群雄         吴国
23    马腾   马超    父亲         群雄         吴国
24    刘备   徐庶    主公         蜀国         吴国
25    曹操   徐庶    主公         魏国         吴国
26    刘胜   刘启    儿子         群雄         吴国
27    刘辩   刘启    同族         群雄         吴国
28    甘氏   刘备     妻         蜀国         吴国
29    刘禅   甘氏    儿子         蜀国         吴国
..   ...  ...   ...        ...        ...
123   曹操   乐进    主公         魏国         吴国
124   曹操   张郃    主公         魏国         吴国
125   曹操   许褚    主公         魏国         吴国
126   曹操   典韦    主公         魏国         吴国
127   曹操   荀彧    主公         魏国         吴国
128   曹操   荀攸    主公         魏国         吴国
129   荀彧   荀攸    叔叔         魏国         吴国
130   曹操   贾诩    主公         魏国         吴国
131   曹操  司马懿    主公         魏国         吴国
132   曹操   程昱    主公         魏国         吴国
133   曹操   于禁    主公         魏国         吴国
134   曹操   邓艾    主公         魏国         吴国
135   曹操   钟会    主公         魏国         吴国
136   曹操   庞德    主公         魏国         吴国
137  司马师  司马懿    儿子         魏国         吴国
138  司马昭  司马师    弟弟         魏国         吴国
139  司马昭  司马懿    儿子         魏国         吴国
140  司马炎  司马昭    儿子         魏国         吴国
141   曹仁   曹操    同族         魏国         吴国
142   曹纯   曹仁    弟弟         魏国         吴国
143   曹昂   曹操    儿子         魏国         吴国
144   刘氏   曹操    妻子         魏国         吴国
145   曹昂   刘氏    儿子         魏国         吴国
146   卞氏   曹操    妻子         群雄         吴国
147   曹丕   卞氏    儿子         吴国         吴国
148   曹植   曹丕    弟弟         魏国         吴国
149   马谡  诸葛亮    同事         蜀国         吴国
150   马谡   刘备     臣         蜀国         吴国
151   孙坚   孙权    父亲         吴国         吴国
152  吴国太   孙权    母亲         吴国         吴国

[153 rows x 5 columns]

保存结果：

In [154]:
path = '../raw_data/triples_processed.txt'
df.to_csv(path, header=None, index=None, sep=',')

输出所有的关系，之后针对这些关系设置一些近义词方便检索

In [30]:
set(df.label.values)

{'主公',
 '义兄',
 '义女',
 '义子',
 '义弟',
 '义父',
 '儿子',
 '叔叔',
 '同事',
 '同族',
 '哥哥',
 '堂弟',
 '夫',
 '女儿',
 '女婿',
 '妹妹',
 '妻',
 '妻子',
 '妾',
 '嫂子',
 '岳母',
 '师傅',
 '弟弟',
 '徒弟',
 '朋友',
 '母亲',
 '父亲',
 '结拜',
 '臣',
 '让位',
 '诸侯'}